src/lib/homestore/shard_manager.cpp

```diff
@@ -20,4 +124,74 @@ ShardManager::Result< InfoList > HSHomeObject::_list_shards(pg_id pg) const {
     return folly::makeUnexpected(ShardError::UNKNOWN_PG);
 }
 
+void HSHomeObject::on_shard_message_commit(int64_t lsn, sisl::blob const& header, sisl::blob const& key,
+                                           void* user_ctx) {
+    const ReplicationMessageHeader* msg_header = r_cast<const ReplicationMessageHeader*>(header.bytes);
+    folly::Promise<ShardManager::Result<ShardInfo> > *promise = nullptr;
+    // user_ctx will be nullptr when:
+    // 1. on the follower side
+    // 2. on the leader side but homeobject restarts and replay all commited log entries from the last commit id;
+    if (user_ctx != nullptr) {
+        promise = r_cast<folly::Promise<ShardManager::Result<ShardInfo> >*> (user_ctx);      
+    }
+
+    if (msg_header->header_crc != msg_header->calculate_crc()) {
+        LOGWARN("replication message header is corrupted with crc error, lsn:{}", lsn);
+        if (promise) {
+            promise->setValue(folly::makeUnexpected(ShardError::INVALID_ARG));	  
+        }
+        return;
+    }
+
+    std::string string_value;
+    string_value.append(r_cast<char*>(header.bytes + sizeof(ReplicationMessageHeader)), msg_header->payload_size);
+
+    auto crc = crc32_ieee(init_crc32, r_cast< const uint8_t* >(string_value.c_str()), string_value.size());
+    if (msg_header->payload_crc != crc) {
+        LOGWARN("replication message body is corrupted with crc error, lsn:{}", lsn);
+        if (promise) {
+            promise->setValue(folly::makeUnexpected(ShardError::INVALID_ARG));	  
+        }
+        return;
+    }
+
+    //TODO: preapre ShardInfoExt msg with chunk_id and persisted the ShardInfoExt to homestore MetaBlkService;
+    void* cookie = nullptr;
+    homestore::hs()->meta_service().add_sub_sb(HomeObjectImpl::s_shard_info_sub_type, r_cast<const uint8_t*>(string_value.c_str()), string_value.size(), cookie);
+
+    //update in-memory shard map;
+    ShardInfoPtr shard_info = deserialize_shard_info(string_value);
+    if (shard_info->state == ShardInfo::State::OPEN) {
+        //create shard;
+        do_commit_new_shard(shard_info);
+    } else {
+        do_commit_seal_shard(shard_info);
+    }
+
+    if (promise) {
+        promise->setValue(ShardManager::Result<ShardInfo>(*shard_info));
+    }
+}
+
+void HSHomeObject::do_commit_new_shard(ShardInfoPtr shard_info) {
+    std::scoped_lock lock_guard(_pg_lock, _shard_lock);  
+    auto pg_iter = _pg_map.find(shard_info->placement_group);
+    RELEASE_ASSERT(pg_iter != _pg_map.end(), "Missing PG info");
+    pg_iter->second.shards.push_back(shard_info);
+
+    //following part is must for follower members or when member is restarted;
+    auto sequence_num = get_sequence_num_from_shard_id(shard_info->id);
```

Comment:
1.follower will happens,  shard is generated from the leader side and  sequence num s increased only in leader side,  this gives the chance for follower to catch up sequence num with leader.
2. it is better to flush pg_sequence_number to disk but it will cause extra IO and I am trying to avoid it by recovering the lastest sequence num from meta blks, as all commited  shard creation is already persisted in meta blks.  there are still another two options for this from my side:   using raft LSN or  combined with (raft term , local sequence num) to be new proposal shard id

Reply:
i prefer  lastest sequence num from meta blks, that is the source of truth

Reply:
that is the current implementation behaviors.